### ISSUE TITLE
Update form Nextcloudvm 17.0 to 17.1 via update.sh deletes redis and went wrong

### ISSUE THREAD

**Issue by horvan** [nextcloud/vm#982]
```bash
joachim@vmd44994:/var/scripts$ sudo bash update.sh
mysql-common set on hold.
mysql-client set on hold.
mysql-client-5.7 set on hold.
mysql-client-core-5.7 set on hold.
mysql-server set on hold.
mysql-server-5.7 set on hold.
mysql-server-core-5.7 set on hold.
mysql-sandbox set on hold.
mysql-source-5.7 set on hold.
mysql-testsuite set on hold.
mysql-testsuite-5.7 set on hold.
mysql-utilities set on hold.
mysql-workbench set on hold.
mysql-workbench-data set on hold.
mysqltcl set on hold.
mysqltuner set on hold.
mariadb-common set on hold.
mariadb-server set on hold.
mariadb-client set on hold.
mariadb-client-10.1 set on hold.
mariadb-client-core-10.1 set on hold.
mariadb-plugin-connect set on hold.
mariadb-server-10.1 set on hold.
mariadb-plugin-cracklib-password-check set on hold.
mariadb-plugin-gssapi-client set on hold.
mariadb-plugin-gssapi-server set on hold.
mariadb-plugin-mroonga set on hold.
mariadb-plugin-oqgraph set on hold.
mariadb-plugin-spider set on hold.
mariadb-plugin-tokudb set on hold.
mariadb-server-core-10.1 set on hold.
mariadb-test set on hold.
mariadb-test-data set on hold.
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
mysql-client was already not hold.
mysql-client-5.7 was already not hold.
mysql-client-core-5.7 was already not hold.
mysql-server was already not hold.
mysql-server-5.7 was already not hold.
mysql-server-core-5.7 was already not hold.
mysql-sandbox was already not hold.
mysql-source-5.7 was already not hold.
mysql-testsuite was already not hold.
mysql-testsuite-5.7 was already not hold.
mysql-utilities was already not hold.
mysql-workbench was already not hold.
mysql-workbench-data was already not hold.
mysqltcl was already not hold.
mysqltuner was already not hold.
Canceled hold on mysql-common.
If you want to upgrade MySQL/MariaDB, please run 'sudo apt update && sudo apt dist-upgrade -y'
mariadb-client was already not hold.
mariadb-plugin-connect was already not hold.
mariadb-plugin-cracklib-password-check was already not hold.
mariadb-plugin-gssapi-client was already not hold.
mariadb-plugin-gssapi-server was already not hold.
mariadb-plugin-mroonga was already not hold.
mariadb-plugin-oqgraph was already not hold.
mariadb-plugin-spider was already not hold.
mariadb-plugin-tokudb was already not hold.
mariadb-test was already not hold.
mariadb-test-data was already not hold.
Canceled hold on mariadb-common.
Canceled hold on mariadb-server.
Canceled hold on mariadb-client-10.1.
Canceled hold on mariadb-client-core-10.1.
Canceled hold on mariadb-server-10.1.
Canceled hold on mariadb-server-core-10.1.
If you want to upgrade MariaDB, please run 'sudo apt update && sudo apt dist-upgrade -y'
Trying to upgrade the Redis PECL extension...
Updating channel "pecl.php.net"
Channel "pecl.php.net" is up to date
Nothing to upgrade
/var/scripts/nextcloud_update.sh: line 180: [: : integer expression expected
Reading package lists... Done
Building dependency tree
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Reading package lists... Done
Building dependency tree
Reading state information... Done
Sourcing file `/etc/default/grub'
Generating grub configuration file ...
Found linux image: /boot/vmlinuz-4.15.0-70-generic
Found initrd image: /boot/initrd.img-4.15.0-70-generic
Found linux image: /boot/vmlinuz-4.15.0-66-generic
Found initrd image: /boot/initrd.img-4.15.0-66-generic
done
dpkg-query: no packages found matching libzfs2linux
/var/scripts/nextcloud_update.sh: line 304: [: : integer expression expected
/dev/fd/63: line 789: [: : integer expression expected
/dev/fd/63: line 825: [: : integer expression expected
An unhandled exception has been thrown:
Doctrine\DBAL\DBALException: Failed to connect to the database: An exception occurred in driver: could not find driver in /var/www/nextcloud/lib/private/DB/Connection.php:64
Stack trace:
#0 /var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(420): OC\DB\Connection->connect()
#1 /var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(380): Doctrine\DBAL\Connection->getDatabasePlatformVersion()
#2 /var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(324): Doctrine\DBAL\Connection->detectDatabasePlatform()
#3 /var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(683): Doctrine\DBAL\Connection->getDatabasePlatform()
#4 /var/www/nextcloud/lib/private/DB/Connection.php(151): Doctrine\DBAL\Connection->setTransactionIsolation(2)
#5 /var/www/nextcloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DriverManager.php(195): OC\DB\Connection->__construct(Array, Object(Doctrine\DBAL\Driver\PDOPgSql\Driver), Object(Doctrine\DBAL\Configuration), Object(Doctrine\Common\EventManager))
#6 /var/www/nextcloud/lib/private/DB/ConnectionFactory.php(156): Doctrine\DBAL\DriverManager::getConnection(Array, Object(Doctrine\DBAL\Configuration), Object(Doctrine\Common\EventManager))
#7 /var/www/nextcloud/lib/private/Server.php(635): OC\DB\ConnectionFactory->getConnection('pgsql', Array)
#8 /var/www/nextcloud/3rdparty/pimple/pimple/src/Pimple/Container.php(118): OC\Server->OC\{closure}(Object(OC\Server))
#9 /var/www/nextcloud/lib/private/ServerContainer.php(120): Pimple\Container->offsetGet('OCP\\IDBConnecti...')
#10 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(161): OC\ServerContainer->query('OCP\\IDBConnecti...')
#11 /var/www/nextcloud/3rdparty/pimple/pimple/src/Pimple/Container.php(114): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(OC\Server))
#12 /var/www/nextcloud/lib/private/ServerContainer.php(120): Pimple\Container->offsetGet('DatabaseConnect...')
#13 /var/www/nextcloud/lib/private/Server.php(1554): OC\ServerContainer->query('DatabaseConnect...')
#14 /var/www/nextcloud/lib/private/Server.php(362): OC\Server->getDatabaseConnection()
#15 /var/www/nextcloud/3rdparty/pimple/pimple/src/Pimple/Container.php(118): OC\Server->OC\{closure}(Object(OC\Server))
#16 /var/www/nextcloud/lib/private/ServerContainer.php(120): Pimple\Container->offsetGet('OC\\Authenticati...')
#17 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(69): OC\ServerContainer->query('OC\\Authenticati...', true)
#18 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(99): OC\AppFramework\Utility\SimpleContainer->buildClass(Object(ReflectionClass))
#19 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(114): OC\AppFramework\Utility\SimpleContainer->resolve('OC\\Authenticati...')
#20 /var/www/nextcloud/lib/private/ServerContainer.php(145): OC\AppFramework\Utility\SimpleContainer->query('OC\\Authenticati...', true)
#21 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(69): OC\ServerContainer->query('OC\\Authenticati...', true)
#22 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(99): OC\AppFramework\Utility\SimpleContainer->buildClass(Object(ReflectionClass))
#23 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(114): OC\AppFramework\Utility\SimpleContainer->resolve('OC\\Authenticati...')
#24 /var/www/nextcloud/lib/private/ServerContainer.php(145): OC\AppFramework\Utility\SimpleContainer->query('OC\\Authenticati...', true)
#25 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(161): OC\ServerContainer->query('OC\\Authenticati...')
#26 /var/www/nextcloud/3rdparty/pimple/pimple/src/Pimple/Container.php(114): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(OC\Server))
#27 /var/www/nextcloud/lib/private/ServerContainer.php(120): Pimple\Container->offsetGet('OC\\Authenticati...')
#28 /var/www/nextcloud/lib/private/Server.php(374): OC\ServerContainer->query('OC\\Authenticati...')
#29 /var/www/nextcloud/3rdparty/pimple/pimple/src/Pimple/Container.php(118): OC\Server->OC\{closure}(Object(OC\Server))
#30 /var/www/nextcloud/lib/private/ServerContainer.php(120): Pimple\Container->offsetGet('OC\\User\\Session')
#31 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(161): OC\ServerContainer->query('OC\\User\\Session')
#32 /var/www/nextcloud/3rdparty/pimple/pimple/src/Pimple/Container.php(114): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(OC\Server))
#33 /var/www/nextcloud/lib/private/ServerContainer.php(120): Pimple\Container->offsetGet('UserSession')
#34 /var/www/nextcloud/lib/private/Server.php(1430): OC\ServerContainer->query('UserSession')
#35 /var/www/nextcloud/lib/private/Server.php(688): OC\Server->getUserSession()
#36 /var/www/nextcloud/3rdparty/pimple/pimple/src/Pimple/Container.php(118): OC\Server->OC\{closure}(Object(OC\Server))
#37 /var/www/nextcloud/lib/private/ServerContainer.php(120): Pimple\Container->offsetGet('OC\\App\\AppManag...')
#38 /var/www/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php(161): OC\ServerContainer->query('OC\\App\\AppManag...')
#39 /var/www/nextcloud/3rdparty/pimple/pimple/src/Pimple/Container.php(114): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(OC\Server))
#40 /var/www/nextcloud/lib/private/ServerContainer.php(120): Pimple\Container->offsetGet('AppManager')
#41 /var/www/nextcloud/lib/private/Server.php(1725): OC\ServerContainer->query('AppManager')
#42 /var/www/nextcloud/lib/private/legacy/app.php(341): OC\Server->getAppManager()
#43 /var/www/nextcloud/lib/private/legacy/app.php(113): OC_App::getEnabledApps()
#44 /var/www/nextcloud/lib/base.php(640): OC_App::loadApps(Array)
#45 /var/www/nextcloud/lib/base.php(1081): OC::init()
#46 /var/www/nextcloud/console.php(46): require_once('/var/www/nextcl...')
#47 /var/www/nextcloud/occ(11): require_once('/var/www/nextcl...')
#48 {main}Sorry but something went wrong. Please report this issue to https://github.com/nextcloud/vm/issues and include the output of the error message. Thank you!
sudo -u www-data php /var/www/nextcloud/occ -V failed
```

**Comment by horvan** [comment]
Ok debug is done

**Comment by horvan** [comment]
I have php 7.2 7.3 and 7.4 Nextcloud uses 7.2 so the update script will fail because the wron php version is set You need to add 
update-alternatives --set php /usr/bin/php7.2 before starting occ because otherwise teh script will fail if more than one phph-version is installed

Thats bad if you choose auto update option because it´'ll ruib redis-server or redis-php package when it runs the first time after installation on ubunrtu 18.04 

I am not sure why this is. But  this does fix it

**Comment by enoch85** [comment]
@horvan Thanks!

The VM comes installed with one version, so that was my only test case. 

How would the script know pre-hand which version to use if there are several installed? Keen to make a PR? :)

**Comment by enoch85** [comment]
@horvan could you please test the latest fix?

**Comment by horvan** [comment]
Should work. I'll test this an report the result here. Maybe it's needed to expose PHPver to reset the update-alternatives configuration, just before the script exits to revoke the system change. 

Maybe there are more bug reports to come. I found out, that the update script ruins Redis-server installation on ubuntu 18.04 currently. I'll take a look.

I'm happy right now because I know that this is the fault of the automatic update script and corresponding cron job. So it's not mine. 

I'll just respawn a server and test the fix as well as trying to find out why redis gets broken when the update script runs the first time after a fresh install. I hope the fix is all needed to solve it.

**Comment by enoch85** [comment]
> I'm happy right now because I know that this is the fault of the automatic update script and corresponding cron job. So it's not mine.

I'm happy for you. :)

I test the scripts regularly by installing it on different servers in different environments, and I've never seen this before. 

As I mentioned before, the VM is not designed to hold several PHP versions, hence if then end user do its own changes then that's on the end user. Of course we could improve, but someone needs to maintain it. Adding more and more features for edge cases are not feasible in the long run IMHO. I would also like to add that the script aren't supposed to be run on installations that aren't VM installations, in case that's the case here?

Regarding redis, could you please provide more information? I did an update yesterday on a server that were on NC 16.0.2 and redis went just fine. 

Thanks!

**Comment by horvan** [comment]
bet that's an ubuntu package issue. because the php-redis module isn't installed after the update. automatically. I am not sure by now if this is just the script and it can be fixed with a simple if then else check. But if this is true that would be very easy to fix. The server is up and running now. I've just to prepare the environment to fire the test case and provide some more information on this.

**Comment by enoch85** [comment]
Here's the current function for installing redis: https://github.com/nextcloud/vm/blob/master/nextcloud_update.sh#L103-L177

**Comment by enoch85** [comment]
Btw, we don't use `php-redis` we use PECL.

**Comment by enoch85** [comment]
And here's the log from an update I just did: https://0bin.net/paste/iNE5yOVTBShjsWbl#mksX5OvZEjMFM+pnt0SYPN6b75gFu1hr1IMEF4WYOdm

No issues with redis or the occ command.

**Comment by enoch85** [comment]
And here's the log from an update I just did: https://0bin.net/paste/Um2N45gt7HK+bVRA#C00n4St7V0bLZY+CS+d5RsmFQ1n5xF1NAaqFO-OxCai

No issues with redis or the occ command.

**Comment by enoch85** [comment]
@horvan Any update here?

I'm leaning towards closing as it works as intended when running the VM as intended, with *one* PHP version installed.

**Comment by horvan** [comment]
ok
